cpp/src/arrow/c/abi.h

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ARROW_FLAG_DICTIONARY_ORDERED 1
+#define ARROW_FLAG_NULLABLE 2
+#define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+struct ArrowSchema {
```

Comment:
I'm struggling with whether or not there should be some concept of next() here and in ArrowArray. It might be helpful to include an example c program that does something with this api using a chunked set of arrays to see whether we're missing part of the functionality here.

Reply:
We don't have a declaration of format version. Are you thinking of it as implicit? How do you think we'll handle iterations? We're just going to introduce an abi2.h or did you have something else in mind?

Reply:
> We're just going to introduce an abi2.h

Yes, this.

With a subtlety: if we merely add new types to the format string or new bits to the flags field, then a new version isn't necessary.

Reply:
> I'm struggling with whether or not there should be some concept of next() here and in ArrowArray

I don't think so. IMHO, any concept of iteration will be application-specific, not part of this interface.

Reply:
> I don't think so. IMHO, any concept of iteration will be application-specific, not part of this interface.

This is also my intuition -- if an application wished to define a `int next(struct ArrowArray*)` callback-type C API this would be straightforward to do and shouldn't impact the array/record batch data structure itself

Reply:
It seems like this would always be necessary in cases where two applications are communicating chunked datasets and we should always think of arrow as a collection of chunked datasets. Leaving this to be different for everyone who is communicating arrow datasets via the c api seems like it is again under-supporting chunked usage patterns.

Reply:
Maybe to put this another way, do you have example integrations in mind which are chunked and don't require basic iteration (or, alternatively, require drastically different implementations of basic iteration)?

Reply:
I think we're talking past each other a little bit. 

I'd be OK with including a C `typedef` defining the expected iterator function pointer interface. I don't think that this function pointer should live as part of either the schema or array data structures that are already defined

Reply:
> I think we're talking past each other a little bit.

Ok. It seems like you want to use the abi for one thing and I want to use it for something slightly different. The boundary of whether we should include schema in the definition is akin to the discussion of whether iteration should be here. There are subjective arguments for each. We already have flatbuf for schema so some could argue that we shouldn't be creating a second encoding. However, it seems useful for the interface. I'd suggest the same is true of the iteration function being on the array so you can get the next array from the current one (a pseudo linked list of arrays/chunk if you will).

Reply:
I'm not arguing with the use case, it's really a question of how to layer the data structures. For example, one possibility is to have a context-like object like:

```
typedef struct ArrowStream {
  struct ArrowArray current_chunk;
  int (*next)(struct ArrowStream*)
};
```

and so the iteration pattern looks like:

```
struct ArrowStream stream;
CHECK_OK(get_stream(..., &stream));
while (stream->next(&stream)) {
  CHECK_OK(process_chunk(&stream->current_chunk));
}
```

Another possibility is to have simply

```
typedef int (*arrow_iterator_t)(struct ArrayArray* next_chunk);

arrow_iterator_t get_next;
CHECK_OK(get_stream(..., &get_next));

struct ArrowArray chunk;
while (get_next(&chunk)) {
  CHECK_OK(process_chunk(&chunk));
}
```

It seems a bit potato-potahto to me. @pitrou will likely have stronger opinions about the C API but I agree that it's a use case that we should support robustly

Reply:
Well, the problem is that you get into the business of designing an API for the application.
* is the iterator sync or async (in other words, does the consumer get the next chunk through a blocking call, or does the provider call a consumer-provided callback when the next chunk is ready)?
* how are error situations signaled (for example if the provider is a Parquet dataset reader or a Flight client)?

The Arrow format is a data format and I'm a bit surprised that we would want to standardize a cross-language programming API for it.